src/bl.cpp

```diff
@@ -622,770 +662,790 @@ static https_request_err_e downloadAndShow()
+        display_show_image(imagePointer, image_reverse, isPNG);
 
-              image_url.toCharArray(filename, image_url.length() + 1);
-              // check if plugin is applied
-              bool flag = preferences.getBool(PREFERENCES_DEVICE_REGISTERED_KEY, false);
-              Log.info("%s [%d]: flag: %d\r\n", __FILE__, __LINE__, flag);
+        // Using filename from API response
+        String new_filename = apiResponse.filename;
 
-              if (apiResponse.filename == "empty_state")
-              {
-                Log.info("%s [%d]: End with empty_state\r\n", __FILE__, __LINE__);
-                if (!flag)
-                {
-                  // draw received logo
-                  status = true;
-                  // set flag to true
-                  if (preferences.getBool(PREFERENCES_DEVICE_REGISTERED_KEY, false) != true) // check the flag to avoid the re-writing
-                  {
-                    bool res = preferences.putBool(PREFERENCES_DEVICE_REGISTERED_KEY, true);
-                    if (res)
-                      Log.info("%s [%d]: Flag written true successfully\r\n", __FILE__, __LINE__);
-                    else
-                      Log.error("%s [%d]: FLag writing failed\r\n", __FILE__, __LINE__);
-                  }
-                }
-                else
-                {
-                  // don't draw received logo
-                  status = false;
-                }
-              }
+        // Print the extracted string
+        Log.info("%s [%d]: New filename - %s\r\n", __FILE__, __LINE__, new_filename.c_str());
+
+        bool res = saveCurrentFileName(new_filename);
+        if (res)
+          Log.info("%s [%d]: New filename saved\r\n", __FILE__, __LINE__);
+        else
+          Log.error("%s [%d]: New image name saving error!", __FILE__, __LINE__);
+
+        if (result != HTTPS_PLUGIN_NOT_ATTACHED)
+          result = HTTPS_SUCCESS;
+      }
+      break;
+      case BMP_FORMAT_ERROR:
+      {
+        error = "First two header bytes are invalid!";
+      }
+      break;
+      case BMP_BAD_SIZE:
+      {
+        error = "BMP width, height or size are invalid";
+      }
+      break;
+      case BMP_COLOR_SCHEME_FAILED:
+      {
+        error = "BMP color scheme is invalid";
+      }
+      break;
+      case BMP_INVALID_OFFSET:
+      {
+        error = "BMP header offset is invalid";
+      }
+      break;
+      default:
+        break;
+      }
+
+      if (isPNG && png_res != PNG_NO_ERR)
+      {
+        filesystem_file_delete("/current.png");
+        submit_log("error parsing image file - %s", error.c_str());
+
+        return HTTPS_WRONG_IMAGE_FORMAT;
+      }
+    }
+  }
+
+  if (send_log)
+  {
+    send_log = false;
+  }
+  Log.info("%s [%d]: Returned result - %d\r\n", __FILE__, __LINE__, result);
+  return result;
+}
+
+https_request_err_e handleApiDisplayResponse(ApiDisplayResponse &apiResponse)
+{
+  https_request_err_e result = HTTPS_NO_ERR;
+
```

Comment:
All the code below is pasted directly from downloadAndShow. It was formatted by my editor but the logic is unchanged.